Inductor: fall back bernoulli on cpu

blzheng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97002


data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

blzheng committed:
Inductor: fall back bernoulli on cpu

[ghstack-poisoned]

blzheng committed:
Update on "Inductor: fall back bernoulli on cpu"



data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/decomposition.py:
```diff
@@ -444,16 +444,16 @@ def view_copy_dtype(self, dtype):
+    @register_extra_random_decomp([aten.bernoulli_])
+    def bernoulli_(self, p=0.5):
+        return self.copy_(torch.rand_like(self, dtype=torch.float32) < p)
 
-
-@register_extra_random_decomp([aten.bernoulli.p])
-def bernoulli_p(self, p=0.5, *, generator=None):
-    assert generator is None
-    return torch.rand_like(self, dtype=torch.float32) < p
+    @register_extra_random_decomp([aten.bernoulli.p])
+    def bernoulli_p(self, p=0.5, *, generator=None):
+        assert generator is None
+        return torch.rand_like(self, dtype=torch.float32) < p
```
It is possible both cpu and gpu are available while we only want to do fallback on cpu. Suggest to check the device type of `self` and raise `NotImplemented` to go directly to lowering if the device is cpu.

blzheng replied:
Thanks for your suggestion. I have updated all checks.

jgong5 commented on torch/_inductor/lowering.py:
```diff
@@ -1059,8 +1059,8 @@ def native_dropout(x, p, train):
     assert (
-        config.fallback_random
-    ), "this should be handled in decomps unless config.fallback_random"
+        config.fallback_random or not has_triton()
+    ), "this should be handled in decomps unless config.fallback_random or not has_triton()"
```
ditto: revise the checks.

blzheng committed:
Update on "Inductor: fall back bernoulli on cpu"



data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

blzheng committed:
Update on "Inductor: fall back bernoulli on cpu"



data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

blzheng committed:
Update on "Inductor: fall back bernoulli on cpu"



data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/decomposition.py:
```diff
@@ -447,11 +447,15 @@ def view_copy_dtype(self, dtype):
 @register_extra_random_decomp([aten.bernoulli_])
 def bernoulli_(self, p=0.5):
+    if self.device == torch.device("cpu"):
+        return NotImplemented
```
```suggestion
        raise NotImplemented
```

blzheng replied:
Its behavior is not as expected if we replace return with raise. I think we can align it with others in this file.
https://github.com/pytorch/pytorch/blob/ee6b19bd4c4f33590697f45dbd492e77dd9b3e34/torch/_inductor/decomposition.py#L249

jgong5 replied:
Oh, makes sense.

jgong5 commented on torch/_inductor/decomposition.py:
```diff
@@ -447,11 +447,15 @@ def view_copy_dtype(self, dtype):
 @register_extra_random_decomp([aten.bernoulli.p])
 def bernoulli_p(self, p=0.5, *, generator=None):
+    if self.device == torch.device("cpu"):
+        return NotImplemented
```
ditto

jgong5 commented on torch/_inductor/overrides.py:
```diff
@@ -35,21 +35,25 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
             and replacements[func] in replacements_using_triton_random
         ):
-            return replacements[func](*args, **kwargs)
+            if func != torch.nn.functional.dropout or (
+                func == torch.nn.functional.dropout
+                and args[0].device != torch.device("cpu")
+            ):
```
How about create a set to hold all the functions to fallback on cpu, e.g., `fallbacks_cpu_random`? Then, we check whether `func` is in it or not here without explicitly checking a specific function here.

blzheng replied:
Done in [7aa97a7](https://github.com/pytorch/pytorch/pull/97002/commits/7aa97a7d8a8bcc3dd7cf40eaad52223cd3eebc45)

jgong5 commented on torch/_inductor/overrides.py:
```diff
@@ -35,21 +35,25 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
                 config.fallback_random
-                and replacements[node.target] in replacements_using_triton_random
-            ):
+                or example_inputs[0].device == torch.device("cpu")
```
safer to check if all example_inputs are on cpu? similar to `fuse_fx`
```python
def fuse_fx(gm: torch.fx.GraphModule, example_inputs):
    is_cpu = all(
        example_input.device == torch.device("cpu")
        for example_input in example_inputs
        if isinstance(example_input, torch.Tensor)
    )
```

blzheng replied:
Changed in [7aa97a7](https://github.com/pytorch/pytorch/pull/97002/commits/7aa97a7d8a8bcc3dd7cf40eaad52223cd3eebc45)

jgong5 commented on torch/_inductor/overrides.py:
```diff
@@ -35,21 +35,25 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
-                and replacements[node.target] in replacements_using_triton_random
-            ):
+                or example_inputs[0].device == torch.device("cpu")
+            ) and replacements[node.target] in replacements_using_triton_random:
```
Do we need to check if it is "dropout"?

blzheng replied:
For now, yes. We verified that aten::bernoulli with MKL optimizations performs much better than rand_like in inductor, but did not observe that rand_like eager implementation outperforming inductor.

blzheng committed:
Update on "Inductor: fall back bernoulli on cpu"



data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/overrides.py:
```diff
@@ -51,6 +59,12 @@ def replace_fx(gm: torch.fx.GraphModule):
                 and replacements[node.target] in replacements_using_triton_random
             ):
                 continue
+            if node.target in fallback_cpu_random and all(
+                example_input.device == torch.device("cpu")
+                for example_input in example_inputs
+                if isinstance(example_input, torch.Tensor)
```
can you lift this cpu device check outside the loop just like `fuse_fx` does? we don't have to compute it multiple times in the loop.

jgong5 commented on torch/_inductor/overrides.py:
```diff
@@ -35,14 +35,22 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
         ):
-            return replacements[func](*args, **kwargs)
+            if func not in fallback_cpu_random or (
+                func in fallback_cpu_random
```
you don't have to check it again here? also, how about combining this if check with the one at line 34?

blzheng committed:
Update on "Inductor: fall back bernoulli on cpu"



data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_inductor/decomposition.py:
```diff
@@ -447,11 +447,15 @@ def view_copy_dtype(self, dtype):
 
 @register_extra_random_decomp([aten.bernoulli_])
 def bernoulli_(self, p=0.5):
+    if self.device == torch.device("cpu"):
```
I don't believe this will be properly marked as a mutation op, so there could be a correctness issue.

blzheng replied:
Sorry, I did not get your point. As for bernoulli_, there is a ut that is passed when fallback_random=True, and the changes here have the same effect as fallback_random=True. Inplace bernoulli on CPU will be handled by [bernoulli_ in lowering](https://github.com/pytorch/pytorch/pull/97002/files#diff-a1b077971cddfabfa0071c5162265066e867bc07721816d95b9cbe58431c38e3L1059-L1067), which can handle mutation properly.https://github.com/pytorch/pytorch/blob/b507d7d798a17d6297871545db0498f6d1d9af07/test/inductor/test_torchinductor.py#L4531-L4542

blzheng commented:
@pytorchbot merge

blzheng commented:
@pytorchbot merge

blzheng committed:
Update on "Inductor: fall back bernoulli on cpu"



data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

blzheng commented:
@pytorchbot merge
